rclcpp_lifecycle/src/lifecycle_node_interface_impl.hpp

```diff
@@ -0,0 +1,401 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIFECYCLE_NODE_INTERFACE_IMPL_HPP_
+#define LIFECYCLE_NODE_INTERFACE_IMPL_HPP_
+
+#include "rclcpp_lifecycle/lifecycle_node.hpp"
+
+#include <functional>
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+#include <utility>
+
+#include "lifecycle_msgs/msg/transition_description.hpp"
+#include "lifecycle_msgs/msg/transition_event.h"  // for getting the c-typesupport
+#include "lifecycle_msgs/msg/transition_event.hpp"
+#include "lifecycle_msgs/srv/change_state.hpp"
+#include "lifecycle_msgs/srv/get_state.hpp"
+#include "lifecycle_msgs/srv/get_available_states.hpp"
+#include "lifecycle_msgs/srv/get_available_transitions.hpp"
+
+#include "rcl/error_handling.h"
+
+#include "rcl_lifecycle/rcl_lifecycle.h"
+
+#include "rclcpp/node_interfaces/node_base_interface.hpp"
+#include "rclcpp/node_interfaces/node_services_interface.hpp"
+
+namespace rclcpp_lifecycle
+{
+
+class LifecycleNode::LifecycleNodeInterfaceImpl
+{
+  using ChangeStateSrv = lifecycle_msgs::srv::ChangeState;
+  using GetStateSrv = lifecycle_msgs::srv::GetState;
+  using GetAvailableStatesSrv = lifecycle_msgs::srv::GetAvailableStates;
+  using GetAvailableTransitionsSrv = lifecycle_msgs::srv::GetAvailableTransitions;
+  using TransitionEventMsg = lifecycle_msgs::msg::TransitionEvent;
+
+public:
+  LifecycleNodeInterfaceImpl(
+    std::shared_ptr<rclcpp::node_interfaces::NodeBaseInterface> node_base_interface,
+    std::shared_ptr<rclcpp::node_interfaces::NodeServicesInterface> node_services_interface)
+  : node_base_interface_(node_base_interface),
+    node_services_interface_(node_services_interface)
+  {}
+
+  ~LifecycleNodeInterfaceImpl()
+  {
+    if (rcl_lifecycle_state_machine_is_initialized(&state_machine_) != RCL_RET_OK) {
+      fprintf(stderr, "%s:%u, FATAL: rcl_state_machine got destroyed externally.\n",
+        __FILE__, __LINE__);
+    } else {
+      rcl_lifecycle_state_machine_fini(&state_machine_,
+        node_base_interface_->get_rcl_node_handle());
+    }
+  }
+
+  void
+  init()
+  {
+    state_machine_ = rcl_lifecycle_get_zero_initialized_state_machine();
+    rcl_ret_t ret = rcl_lifecycle_state_machine_init(
+      &state_machine_, node_base_interface_->get_rcl_node_handle(),
+      ROSIDL_GET_TYPE_SUPPORT(lifecycle_msgs, msg, TransitionEvent),
+      rosidl_typesupport_cpp::get_service_type_support_handle<ChangeStateSrv>(),
+      rosidl_typesupport_cpp::get_service_type_support_handle<GetStateSrv>(),
+      rosidl_typesupport_cpp::get_service_type_support_handle<GetAvailableStatesSrv>(),
+      rosidl_typesupport_cpp::get_service_type_support_handle<GetAvailableTransitionsSrv>(),
```

Comment:
This is neat. 😄

It might be a good idea to explain why both C and C++ type supports are used, so that when read it again in the future it is less confusing.